File: app/src/fizzbuzz.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { FizzBuzz } from './fizzbuzz';

describe('FizzBuzz', () => {
  let fizzbuzz: FizzBuzz;

  beforeEach(() => {
    fizzbuzz = new FizzBuzz();
  });

  describe('数を文字列にして返す', () => {
    describe('タイプ1の場合', () => {
      beforeEach(() => {
        fizzbuzz = new FizzBuzz(1);
      });

      describe('その他の場合', () => {
        it('1を渡したら文字列"1"を返す', () => {
          expect(fizzbuzz.generate(1)).toBe('1');
        });

        it('2を渡したら文字列"2"を返す', () => {
          expect(fizzbuzz.generate(2)).toBe('2');
        });
      });

      describe('3の倍数の場合', () => {
        it('3を渡したら文字列"Fizz"を返す', () => {
          expect(fizzbuzz.generate(3)).toBe('Fizz');
        });

        it('6を渡したら文字列"Fizz"を返す', () => {
          expect(fizzbuzz.generate(6)).toBe('Fizz');
        });
      });

      describe('5の倍数の場合', () => {
        it('5を渡したら文字列"Buzz"を返す', () => {
          expect(fizzbuzz.generate(5)).toBe('Buzz');
        });
      });

      describe('3と5両方の倍数の場合', () => {
        it('15を渡したら文字列"FizzBuzz"を返す', () => {
          expect(fizzbuzz.generate(15)).toBe('FizzBuzz');
        });
      });

      describe('1から100までのFizzBuzzの配列を返す', () => {
        let result: string[];

        beforeEach(() => {
          const fb = new FizzBuzz(1);
          fb.generateList();
          result = fb.list;
        });

        it('配列の初めは文字列の1を返す', () => {
          expect(result[0]).toBe('1');
        });

        it('配列の最後は文字列のBuzzを返す', () => {
          expect(result[99]).toBe('Buzz'); // 100番目の要素
        });

        it('配列の3番目は文字列のFizzを返す', () => {
          expect(result[2]).toBe('Fizz'); // 3番目の要素
        });

        it('配列の5番目は文字列のBuzzを返す', () => {
          expect(result[4]).toBe('Buzz'); // 5番目の要素
        });

        it('配列の15番目は文字列のFizzBuzzを返す', () => {
          expect(result[14]).toBe('FizzBuzz'); // 15番目の要素
        });
      });
    });
  });

  describe('タイプごとに出力を切り替えることができる', () => {
    describe('タイプ1の場合', () => {
      beforeEach(() => {
        fizzbuzz = new FizzBuzz(1);
      });

      it('1を渡したら文字列"1"を返す', () => {
        expect(fizzbuzz.generate(1)).toBe('1');
      });
    });

    describe('タイプ2の場合', () => {
      beforeEach(() => {
        fizzbuzz = new FizzBuzz(2);
      });

      describe('3の倍数の場合', () => {
        it('3を渡したら文字列"3"を返す', () => {
          expect(fizzbuzz.generate(3)).toBe('3');
        });
      });

      describe('5の倍数の場合', () => {
        it('5を渡したら文字列"5"を返す', () => {
          expect(fizzbuzz.generate(5)).toBe('5');
        });
      });

      describe('3と5両方の倍数の場合', () => {
        it('15を渡したら文字列"15"を返す', () => {
          expect(fizzbuzz.generate(15)).toBe('15');
        });
      });

      describe('その他の場合', () => {
        it('1を渡したら文字列"1"を返す', () => {
          expect(fizzbuzz.generate(1)).toBe('1');
        });
      });
    });

    describe('タイプ3の場合', () => {
      beforeEach(() => {
        fizzbuzz = new FizzBuzz(3);
      });

      describe('3の倍数の場合', () => {
        it('3を渡したら文字列"3"を返す', () => {
          expect(fizzbuzz.generate(3)).toBe('3');
        });
      });

      describe('5の倍数の場合', () => {
        it('5を渡したら文字列"5"を返す', () => {
          expect(fizzbuzz.generate(5)).toBe('5');
        });
      });

      describe('3と5両方の倍数の場合', () => {
        it('15を渡したら文字列"FizzBuzz"を返す', () => {
          expect(fizzbuzz.generate(15)).toBe('FizzBuzz');
        });
      });

      describe('その他の場合', () => {
        it('1を渡したら文字列"1"を返す', () => {
          expect(fizzbuzz.generate(1)).toBe('1');
        });
      });
    });

    describe('それ以外のタイプの場合', () => {
      beforeEach(() => {
        fizzbuzz = new FizzBuzz(4);
      });

      it('1を渡したら例外を返す', () => {
        expect(() => fizzbuzz.generate(1)).toThrow('タイプが未指定です');
      });
    });
  });
});

File: app/fizz_buzz/test/fizz_buzz_test.exs
defmodule FizzBuzzTest do
  use ExUnit.Case

  describe "タイプごとに出力を切り替えることができる - タイプ1の場合" do
    test "1を渡したら文字列1を返す" do
      assert FizzBuzz.generate(1, 1) == "1"
    end

    test "2を渡したら文字列2を返す" do
      assert FizzBuzz.generate(2, 1) == "2"
    end

    test "3を渡したら文字列Fizzを返す" do
      assert FizzBuzz.generate(3, 1) == "Fizz"
    end

    test "5を渡したら文字列Buzzを返す" do
      assert FizzBuzz.generate(5, 1) == "Buzz"
    end

    test "15を渡したら文字列FizzBuzzを返す" do
      assert FizzBuzz.generate(15, 1) == "FizzBuzz"
    end

    test "1から100までの数を返す" do
      result = FizzBuzz.list(100)
      assert length(result) == 100
      assert Enum.at(result, 0) == "1"
      assert Enum.at(result, 1) == "2"
      assert Enum.at(result, 99) == "Buzz"
    end

    test "1から100までのFizzBuzzの配列を返す" do
      result = FizzBuzz.list(100)
      # 3
      assert Enum.at(result, 2) == "Fizz"
      # 5
      assert Enum.at(result, 4) == "Buzz"
      # 15
      assert Enum.at(result, 14) == "FizzBuzz"
    end
  end

  describe "タイプごとに出力を切り替えることができる - タイプ2の場合" do
    test "1を渡したら文字列1を返す" do
      assert FizzBuzz.generate(1, 2) == "1"
    end
  end

  describe "タイプごとに出力を切り替えることができる - タイプ3の場合" do
    test "1を渡したら文字列1を返す" do
      assert FizzBuzz.generate(1, 3) == "1"
    end
  end

  describe "タイプごとに出力を切り替えることができる - それ以外のタイプの場合" do
    test "存在しないタイプを渡したらエラーになる" do
      assert_raise CaseClauseError, fn ->
        FizzBuzz.generate(1, 4)
      end
    end
  end
end
